### PR TITLE
Support clr.GetClrType() - as in IronPython

### PR DESCRIPTION
**Pull request by vivainio**
### What does this implement/fix? Explain your changes.

Implements https://github.com/pythonnet/pythonnet/issues/432

### Does this close any currently open issues?

Yes, #432 

### Any other comments?

...

### Checklist

Check all those that are applicable and complete.

-   [x] Make sure to include one or more tests for your change
-   [x] If an enhancement PR, please create docs and at best an example
-   [x] Add yourself to [`AUTHORS`](../blob/master/AUTHORS.md)
-   [x] Updated the [`CHANGELOG`](../blob/master/CHANGELOG.md)

### CODE REVIEW

**Comment by vivainio** [comment]
This was surprisingly easy, as the type of object passed to GetClrType was actually an instance of Type already.

The implementation just passes it back after a cast, yielding a RuntimeType instance that you can do reflection on.

I will document the feature after you agree in principle that this is the right thing to do.

**Comment by denfromufa** [comment]
Can you please add a minimal test?

File: src/tests/test_module.py [hunk 1]
```diff
@@ -352,6 +352,16 @@ def test_clr_add_reference():
     with pytest.raises(FileNotFoundException):
         AddReference("somethingtotallysilly")
 
+def test_clr_get_clr_type():
```
Comment (denfromufa) [inline src/tests/test_module.py]: can you add the test to invoke the ArgumentException error?

Comment (vivainio) [inline src/tests/test_module.py]: Done

**Comment by denfromufa** [comment]
@vivainio this looks great, please add yourself to authors, this feature to changelog, and minimal doc somewhere.

**Comment by filmor** [comment]
Note that this behaviour is not exactly the same as what IronPython is doing: https://github.com/IronLanguages/main/blob/ipy-2.7.7/Languages/IronPython/IronPython/Runtime/ClrModule.cs#L525

I haven't tried this out but it looks like the exception they raise here would be `TypeError` because of the method resolution failing. Couldn't you use the exact same implementation?

**Comment by denfromufa** [comment]
@filmor i just tried and you are right:

```
IronPython 2.7.5 (2.7.5.0) on .NET 4.0.30319.42000 (32-bit)
Type "help", "copyright", "credits" or "license" for more information.
>>> import clr
>>> clr.GetClrType
<built-in function GetClrType>
>>> clr.GetClrType(1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: expected Type, got int
```

**Comment by denfromufa** [comment]
I deleted my previous message, here is how it works in ironpython:

```
IronPython 2.7.5 (2.7.5.0) on .NET 4.0.30319.42000 (32-bit)
Type "help", "copyright", "credits" or "license" for more information.
>>> import clr
>>> clr.GetClrType
<built-in function GetClrType>
>>> clr.GetClrType(1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: expected Type, got int

>>> clr.GetClrType(int)
<System.RuntimeType object at 0x000000000000002B [System.Int32]>
>>> clr.GetClrType(int(1))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: expected Type, got int

>>> type(1)
<type 'int'>

>>> import System
>>> clr.GetClrType(System.Int32(1))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: expected Type, got int

>>>
```

**Comment by denfromufa** [comment]
@vivainio besides the exception type, here is something else I found that needs to be handled like IPY for compatibility - uninitialized reference types. Can you add this to tests? Here is comparison between C# and IPY. Note that C# requires overloaded method TypeName (found it on SO somewhere):

```python
IronPython 2.7.7 (2.7.7.0) on .NET 4.0.30319.42000 (32-bit)
Type "help", "copyright", "credits" or "license" for more information.
>>> import clr
>>> import System
>>> dblarr=System.Array[System.Double]
>>> clr.GetClrType(dblarr)
<System.RuntimeType object at 0x000000000000006C [System.Double[]]>
>>>
```

```csharp
scriptcs (ctrl-c to exit or :help for help)

> double[] dblarr;

> Type TypeName<T>(T obj) { return typeof(T); }
> Type TypeName(object obj) { if (obj != null) { return obj.GetType(); } else {
return null; } }
> TypeName(dblarr)
System.Double[], mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a
5c561934e089
> (Type)dblarr
(1,1): error CS0030: Cannot convert type 'double[]' to 'System.Type'
> typeof(dblarr)
(1,8): error CS0246: The type or namespace name 'dblarr' could not be found (are
 you missing a using directive or an assembly reference?)
> dblarr.GetType()
Object reference not set to an instance of an object.
```

**Comment by vivainio** [comment]
It seems the implementation nicked from IronPython handles all the cases you described. I'll push a new version with that & requisite tests later today

File: src/tests/test_module.py [hunk 2]
```diff
@@ -352,6 +352,26 @@ def test_clr_add_reference():
...
+    assert comparable.FullName == "System.IComparable"
+    assert comparable.IsInterface
+    assert GetClrType(int).FullName == "System.Int32"
+    assert GetClrType(str).FullName == "System.String"
+    assert GetClrType(float).FullName == "System.Double"
+    dblarr = System.Array[System.Double]
+    assert GetClrType(dblarr).FullName == "System.Double[]"
+
+    with pytest.raises(TypeError):
+        GetClrType(1)
```
Comment (denfromufa) [inline src/tests/test_module.py]: can you compare also against types, not instance of types?

Comment (vivainio) [inline src/tests/test_module.py]: Can you elaborate, what kind of comparison I should add?

Comment (denfromufa) [inline src/tests/test_module.py]: GetClrType(int), GetClrType(str)

Comment (vivainio) [inline src/tests/test_module.py]: Check lines 365-367 in the diff :)

Comment (denfromufa) [inline src/tests/test_module.py]: Oh, I see now let's wait for @filmor to approve

**Comment by obulat** [comment]
I'm working on Toga project (pybee), and getting an error when trying to import `GetClrType`:
`ImportError: cannot import name 'GetClrType'`

**Comment by denfromufa** [comment]
@obulat this is in master branch, not on PYPI yet.

**Comment by obulat** [comment]
Oh, thanks for the reply! Do you have any information about when it will be on PYPI?

**Comment by denfromufa** [comment]
Whenever we fix one release blocking issue with numpy usage.

**Pull request merged**